compiler/optimizer/abstractinterpreter/AbsValue.cpp

```diff
@@ -0,0 +1,92 @@
+/*******************************************************************************
+ * Copyright (c) 2020, 2020 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at http://eclipse.org/legal/epl-2.0
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "optimizer/abstractinterpreter/AbsValue.hpp"
+
+TR::AbsValue* TR::AbsVPValue::clone(TR::Region& region)
+   {
+   TR::AbsVPValue* copy = new (region) TR::AbsVPValue(_vp, _constraint, _dataType);
+   copy->setParamPosition(_paramPos);
+
+   if (_isImplicitParameter)
+      copy->setImplicitParam();
+   
+   return copy;
+   }
+
+TR::AbsValue* TR::AbsVPValue::merge(TR::AbsValue *other)
+   {
+   TR_ASSERT_FATAL(other, "Cannot merge with a NULL AbsValue");
+
+   if (other->getDataType() != _dataType) 
+      return NULL;
+
+   if (!_constraint)
+      return this;
+
+   if (_paramPos != other->getParamPosition()) 
+      _paramPos = -1;
+
+   if (_isImplicitParameter && !other->isImplicitParameter())
+      _isImplicitParameter = false;
+
+   TR::AbsVPValue* otherVPValue = static_cast<TR::AbsVPValue*>(other);
+
+   if (otherVPValue->isTop()) 
+      {
+      setToTop();
+      return this;
+      }
+
+   TR::VPConstraint *mergedConstraint = _constraint->merge(otherVPValue->getConstraint(), _vp);
+
+   if (mergedConstraint) //mergedConstaint can be VPMergedIntConstraint or VPMergedLongConstraint. Turn them into VPIntRange or VPLongRange to make things easier.
+      {
+      if (mergedConstraint->asMergedIntConstraints())
+         mergedConstraint = TR::VPIntRange::create(_vp, mergedConstraint->getLowInt(), mergedConstraint->getHighInt());
+      else if (mergedConstraint->asMergedLongConstraints())
+         mergedConstraint = TR::VPLongRange::create(_vp, mergedConstraint->getLowLong(), mergedConstraint->getHighLong());
+      }
```

Comment:
The merge() function of `TR::VPConstraint` can produce `TR::VPMergedConstraint` when merging int or long constraints. For example, 1 merges with 5, we will get a merged constraint {1, 5 }, if the results then merged with another int 10, it becomes {1, 5, 10}. `TR::VPMergedConstraint` is not useful during abstract interpretation because knowing the upper bound and the lower bound of an integer is enough, which is <1, 10>. `TR::VPIntRange` does exactly this job.

Also, if you look at the class hierarchy info of TR::VPMergedConstraint, it inherits directly from TR::VPConstraint. So during abstract interpretation, if not converting a merged constraint to TR::VPIntRange or TR::VPLongRange, we have to check for an additional type, which will be a lot of work. By converting merged constraint to a range constraint, we only need to check if a constraint is an int or long constraint by using _asIntConstraint()_ or `asLongConstraint()`, or more concrete type by using `asIntConst()` or `asIntRange()`

Reply:
TR::VPMergedConstraint is not useful during abstract interpretation, knowing the upper bound and the lower bound of an int or a long value is enough.

if you look at the class hierarchy info of TR::VPMergedConstraint, it inherits directly from TR::VPConstraint. So during abstract interpretation, if not converting a merged constraint to TR::VPIntRange or TR::VPLongRange, we have to check for an additional type, which will be a lot of work. By converting merged constraint to a range constraint, we only need to check if a constraint is an int or long constraint by using _asIntConstraint()_ or `asLongConstraint()`, or more concrete type by using `asIntConst()` or `asIntRange()`